src/main/java/software/amazon/documentdb/jdbc/sshtunnel/DocumentDbSshTunnelServer.java

```diff
@@ -115,69 +107,294 @@ private DocumentDbSshTunnelServer(final DocumentDbSshTunnelServerBuilder builder
+                return;
+            }
+            validateLocalSshFilesExists();
+            session = createSshTunnel(getConnectionProperties());
         }
     }
 
     /**
      * Removes a client from the reference count for this server. If the reference count reaches zero, then
      * the serve attempt to stop the SSH Tunnel service.
      *
-     * @throws Exception When an error occur attempting shutdown of the service process.
+     * @throws SQLException When an error occur attempting shutdown of the service process.
      */
-    public void removeClient() throws Exception {
-        synchronized (MUTEX) {
-            if (clientCount.decrementAndGet() == 0) {
-                close();
+    public void removeClient() throws SQLException {
+        synchronized (mutex) {
+            if (clientCount.get() <= 0 || clientCount.decrementAndGet() > 0) {
+                return;
             }
+            closeSession();
         }
     }
 
-    private void shutdownServerWatcherThread() throws Exception {
-        if (serverWatcherThread.isAlive()) {
-            LOGGER.info("Stopping server watcher thread.");
-            serverWatcher.close();
-            do {
-                serverWatcherThread.join(SERVER_WATCHER_POLL_TIME_MS * 2);
-            } while (serverWatcherThread.isAlive());
-            final Queue<Exception> exceptions = serverWatcher.getExceptions();
-            if (!exceptions.isEmpty()) {
-                for (Exception e : exceptions) {
-                    LOGGER.error(e.getMessage(), e);
+    private void closeSession() throws SQLException {
+        if (scheduledFuture != null) {
+            cancelScheduledFutureClose();
+        }
+        // Delay the close, if indicated.
+        final long delayMS = getCloseDelayMS();
+        if (delayMS <= 0) {
+            close();
+        } else {
+            LOGGER.debug("Close timer is being scheduled.");
+            scheduledFuture = scheduler.schedule(getCloseTimerTask(), delayMS, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private TimerTask getCloseTimerTask() {
+        return new TimerTask() {
+            @Override
+            public void run() {
+                try {
+                    close();
+                } catch (Exception e) {
+                    // Ignore exception on close.
+                    LOGGER.warn(e.getMessage(), e);
                 }
             }
-            LOGGER.info("Stopped server watcher thread.");
-        } else {
-            LOGGER.info("Server watcher thread already stopped.");
+        };
+    }
+
+    private void cancelScheduledFutureClose() throws SQLException {
+        synchronized (mutex) {
+            LOGGER.debug("Close timer is being cancelled.");
+            while (!scheduledFuture.isDone()) {
+                scheduledFuture.cancel(false);
+                try {
+                    TimeUnit.MILLISECONDS.sleep(10);
+                } catch (InterruptedException e) {
```

Comment:
Is it possible to have an exception for a sleep()?

Reply:
Yes. It can be interrupted.